multicluster/controllers/multicluster/label_identity_controller.go

```diff
@@ -154,160 +146,79 @@ func (r *LabelIdentityReconciler) namespaceMapFunc(ns client.Object) []reconcile
 	return requests
 }
 
-// onPodDelete removes the Pod and label identity mapping from the cache, and
-// updates label identity ResourceExport if necessary (the Pod deletion event
-// causes a label identity no longer present in the cluster).
-func (r *LabelIdentityReconciler) onPodDelete(ctx context.Context, podNamespacedName string) error {
-	labelToDelete := r.getLabelToDelete(podNamespacedName)
-	if labelToDelete != "" {
-		if err := r.deleteLabelIdentityResExport(ctx, labelToDelete); err != nil {
-			return err
-		}
-		r.deleteFromLabelCache(podNamespacedName, labelToDelete)
+// onPodCreateOrUpdate handles Pod create or update events.
+func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podNamespacedName, currentLabel string) error {
+	if err := r.checkForLabelResExportDelete(ctx, podNamespacedName, currentLabel); err != nil {
+		return err
 	}
-	return nil
+	return r.checkForLabelResExportAdd(ctx, podNamespacedName, currentLabel)
 }
 
-// onPodCreateOrUpdate updates the Pod and label identity mapping in the cache, and
-// updates label identity ResourceExport if necessary (the Pod creation/update
-// event causes a new label identity to appear in the cluster or a label identity
-// no longer present in the cluster).
-func (r *LabelIdentityReconciler) onPodCreateOrUpdate(ctx context.Context, podNamespacedName, normalizedLabel string) error {
-	labelToAdd, labelToDelete := r.getLabelsToUpdate(podNamespacedName, normalizedLabel)
-	if labelToDelete != "" {
-		if err := r.deleteLabelIdentityResExport(ctx, labelToDelete); err != nil {
-			return err
-		}
-		r.deleteFromLabelCache(podNamespacedName, labelToDelete)
-	}
-	if labelToAdd != "" {
-		if err := r.createLabelIdentityResExport(ctx, labelToAdd); err != nil {
-			return err
+// checkForLabelResExportDelete removes the Pod and label identity mapping from the cache,
+// and deletes label identity ResourceExport if necessary (the Pod deletion event
+// causes a label identity no longer present in the cluster).
+func (r *LabelIdentityReconciler) checkForLabelResExportDelete(ctx context.Context, podNamespacedName, currentLabel string) error {
+	originalLabel, isCached := r.podLabelCache[podNamespacedName]
+	if isCached && originalLabel != currentLabel {
+		unlock := r.labelMutexes.Lock(originalLabel)
+		defer unlock()
+		if podNames, ok := r.labelToPodsCache[originalLabel]; ok && len(podNames) == 1 && podNames.Has(podNamespacedName) {
```

Comment:
Need to protect labelToPodsCache.

Reply:
Same as above, I thought with per label lock this should be ok?

Reply:
No. We need to protect the map itself. You can use sync.Map too.

Reply:
Updated